src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -133,34 +133,97 @@ public ItemData(I item, int elementOrder, bool conditionResult)
             public bool ConditionResult { get { return _conditionResult; } }
         }
 
+        private class OperationWithCache : IItemOperation
```

Comment:
I like MemoizedOperation. It's just that it's not self-memoizing and needs an external caller to tell it when to cache.